kernel/aster-nix/src/net/socket/ip/stream/mod.rs

```diff
@@ -414,6 +392,12 @@ impl FileLike for StreamSocket {
     }
 }
 
+impl Pollable for StreamSocket {
+    fn poll(&self, mask: IoEvents, poller: Option<&Poller>) -> IoEvents {
+        <Self as FileLike>::poll(self, mask, poller)
```

Comment:
`FileLike` provides a dummy implementation of `poll` by default. I don't think it is appropriate to add the `wait_events` method in this case (`wait_events` will become a busy loop if `poll` does not work as normal):
```rust
    fn poll(&self, _mask: IoEvents, _poller: Option<&Poller>) -> IoEvents {
        IoEvents::empty()
    }
```

Since the majority of the `FileLike` implementors do not use `wait_events` either, I do not think it is appropriate to make `FileLike` a sub-trait of `Pollable`, at least for now.

Reply:
But having this `poll` method

```rust
fn poll(&self, _mask: IoEvents, _poller: Option<&Poller>) -> IoEvents
```

in both `FileLike` and `Pollable` traits is confusing as these two versions of `poll` methods share exactly the same semantics. People will scratch their heads about the rationale of keeping two versions of `poll`.

Your concern is that if the `poll` method is implemented _incorrectly_, then `wait_events` will never return. In this case, I think we should blame the incorrect `poll` method, not `wait_events`. And any wait-for-event method (such as `wait_events`) has the inherent semantics of allowing the possibility of blocking indefinitely. To prevent such unfortunate situation, a timeout version of `wait_events` should be provided and used instead.

In fact, I think we should NOT provide the default implementation.

```
    fn poll(&self, _mask: IoEvents, _poller: Option<&Poller>) -> IoEvents {
        IoEvents::empty()
    }
```

This default implementation is almost always wrong as a file should be at least readable or writable sometimes.

Reply:
OK, I am convinced by your argument, especially after getting a little more experience refactoring the UNIX sockets.

A few notes on "a little more experience": I have found that the right place to implement `Pollable` is on `PipeReader`/`PipeWriter` and `UnixStreamSocket`, instead of the `Producer`/`Consumer` in `channel.rs`. There are several reasons for this decision, which I will explain in detail in future PRs. But one thing worth noting is that all these types are `FileLike`, so making `FileLike` a sub-trait makes things even simpler.

So let's make `FileLike` a sub-trait of `Pollable` in this PR.

Reply:
Well, but unfortunately:
```
error[E0038]: the trait `fs::file_handle::FileLike` cannot be made into an object
   --> kernel/aster-nix/src/syscall/write.rs:30:21
    |
30  |     let write_len = file.write(&buffer)?;
    |                     ^^^^^^^^^^^^^^^^^^^ `fs::file_handle::FileLike` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> kernel/aster-nix/src/process/signal/poll.rs:272:8
    |
272 |     fn wait_events<F, R>(&self, mask: IoEvents, mut cond: F) -> Result<R>
    |        ^^^^^^^^^^^ ...because method `wait_events` has generic type parameters
    |
   ::: kernel/aster-nix/src/fs/file_handle.rs:22:11
    |
22  | pub trait FileLike: Pollable + Send + Sync + Any {
    |           -------- this trait cannot be made into an object...
    = help: consider moving `wait_events` to another trait
    = note: `fs::file_handle::FileLike` can be implemented in other crates; if you want to support your users passing their own types here, you can't refer to a specific type
```

So making `FileLike` a sub-trait of `Pollable` won't compile. Here is a minimal example ([Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e18b5e6eee5d231b7faefcad2f2d7ab2)):
```rust
trait A {
    fn work<T>(&self, t: T);
    fn hello(&self);
}

fn foo(x: &dyn A) {
    x.hello();
}

fn main() {
    println!("Hello, world!");
}
```

Reply:
FYI, we can work around this problem by writing something like:

```rust
pub trait Pollable {
    fn poll(&self, _mask: IoEvents, _poller: Option<&Poller>) -> IoEvents;
}

pub trait WaitEvents {
    fn wait_events<F, R>(&self, mask: IoEvents, mut cond: F) -> Result<R>
    where
        F: FnMut() -> Result<R>;
}

impl<T: Pollable> WaitEvents for T {
    // [..]
}
```

I don't have a preference between the current implementation and this one, because I think this one brings both pros (e.g., `FileLike` can be a sub-trait of `Pollable` and `wait_events` cannot be arbitrarily overridden for `T: Pollable`) and cons (e.g., the traits are more complex, since we need two traits `WaitEvents` and `Pollable` instead of one). What do you think?

Reply:
Maybe you don't need to introduce the trait `WaitEvents`, you can just implement `wait_events` on `dyn Pollable`.

```rust
impl dyn Pollable {
    pub fn wait_events(&self) {}
}
```

Reply:
> Maybe you don't need to introduce the trait WaitEvents, you can just implement wait_events on dyn Pollable.

This will force us to write something like `(self as &dyn Pollable).wait_events(..)` instead of `self.wait_events(..)`. IMO this is ugly.

A minimal example ([Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2c90e207b9530c8f2613744cadccdcb4)):
```rust
#![allow(dead_code)]

trait A {
    fn read(&self);
}

impl dyn A {
    fn wait<T>(&self, _t: T) {}
}

impl<T: 'static> A for T {
    fn read(&self) {
        println!("Hello, world!");
        
        // self.wait(1usize); // This won't compile.
        (self as &dyn A).wait(1usize);
    }
}

fn main() {}
```

Reply:
Try this one.

```rust
pub trait Pollable {
    fn poll(&self, _mask: IoEvents, _poller: Option<&Poller>) -> IoEvents;

    fn wait_events<F, R>(&self, mask: IoEvents, mut cond: F) -> Result<R>
    where
        Self: Sized,
        F: FnMut() -> Result<R>;   
}
```

The revised minimal sample code is as follows.

```rust
trait A {
    fn work<T>(&self, t: T)
    where
        Self: Sized;
    fn hello(&self);
}

fn foo(x: &dyn A) {
    x.hello();
}

fn foo2<T: A>(x: T) {
    x.work(0u32);
}

fn main() {
    println!("Hello, world!");
}
```

Reply:
Great! I'll have a try latter.

It's surprising to me that the rust compiler isn't smart enough to suggest this solution, instead suggesting `` consider moving `wait_events` to another trait ``.